eclipse-theia/theia
# Improve readonly editor behavior

#### What it does

Improves our readonly handling for monaco editors (editors can now dynamically change being readonly and writable).

Additionally adds a readonly mode to notebook editors. This new readonly mode disables a lot of notebook features:
* Editing (code & markdown)
* Execution
* Drag & drop
* Creating new cells

This change lays the groundwork for a few features in the collaboration feature (where readonly can be toggled) and the changes required for https://github.com/eclipse-theia/theia/issues/13353.

#### How to test

1. Open a monaco editor (or notebook editor).
2. Run the command `Toggle File System Readonly`. It toggles whether the `file` URI scheme is treated as readonly.
3. Play around with the editors. You should not be able to manipulate the editors.
4. Open a new editor. It should open in readonly mode.
5. Run the command again. Everything should work as normal.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
